api/client/proxy/client_test.go

```diff
@@ -678,3 +680,57 @@ func TestClusterCredentials(t *testing.T) {
 		})
 	}
 }
+
+func TestNewDialerForGRPCClient(t *testing.T) {
+	t.Run("Check that PROXYHeaderGetter if present sends PROXY header as first bytes on the connection", func(t *testing.T) {
+		listener, err := net.Listen("tcp", "127.0.0.1:0")
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			require.NoError(t, listener.Close())
+		})
+
+		prefix := []byte("FAKEPROXY")
+		proxyHeaderGetter := func() ([]byte, error) {
+			return prefix, nil
+		}
+
+		ctx := context.Background()
+		cfg := &ClientConfig{
+			PROXYHeaderGetter: proxyHeaderGetter,
+		}
+		dialer := newDialerForGRPCClient(ctx, cfg)
+
+		resultChan := make(chan bool)
+		// Start listening, emulating receiving end of connection
+		go func() {
+			conn, err := listener.Accept()
+			if err != nil {
+				assert.Fail(t, err.Error())
+				return
+			}
+
+			buf := make([]byte, len(prefix))
+			_, err = conn.Read(buf)
+			assert.NoError(t, err)
+			t.Cleanup(func() {
+				require.NoError(t, conn.Close())
+			})
+
+			// On the received connection first bytes should be our PROXY prefix
+			resultChan <- slices.Equal(buf, prefix)
+		}()
+
+		conn, err := dialer(ctx, listener.Addr().String())
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			require.NoError(t, conn.Close())
+		})
+
+		select {
+		case res := <-resultChan:
+			require.True(t, res, "Didn't receive required prefix as first bytes on the connection")
+		case <-time.After(time.Second):
```

Comment:
One second... sounds optimistic :see_no_evil:

Reply:
Not much happening there, if it takes more to just dial we have some serious problems with latency 😅